File: go/pkg/basecamp/download.go
package basecamp

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"path"
	"time"
)

// fetchSignedDownload fetches content from a signed download URL (e.g., S3).
// Uses the bare transport (no loggingTransport, no auth headers) and no
// client-level timeout so the caller owns the streaming lifecycle.
func (c *Client) fetchSignedDownload(ctx context.Context, downloadURL string) (*http.Response, error) {
	req, err := http.NewRequestWithContext(ctx, "GET", downloadURL, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create download request: %w", err)
	}

	transport := c.httpOpts.Transport
	if transport == nil {
		transport = newDefaultTransport()
	}
	httpClient := &http.Client{
		Transport: transport,
		Timeout:   0, // no client-level timeout — streaming owned by caller
	}

	resp, err := httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to download file: %w", err)
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		_ = resp.Body.Close()
		return nil, ErrAPI(resp.StatusCode, fmt.Sprintf("download failed with status %d", resp.StatusCode))
	}

	return resp, nil
}

// DownloadURL fetches file content from any API-routable download URL.
//
// Handles the full download flow: URL rewriting to the configured API host,
// authenticated first hop (which typically 302s to a signed download URL),
// and unauthenticated second hop to fetch the actual file content. Common
// inputs include storage blob URLs from <bc-attachment> elements and any
// other signed-download URL that routes through the API.
//
// The caller is responsible for closing the returned Body.
func (ac *AccountClient) DownloadURL(ctx context.Context, rawURL string) (result *DownloadResult, err error) {
	// Validation
	if rawURL == "" {
		return nil, ErrUsage("download URL is required")
	}
	parsed, parseErr := url.Parse(rawURL)
	if parseErr != nil || !parsed.IsAbs() {
		return nil, ErrUsage("download URL must be an absolute URL")
	}
	if parsed.Scheme != "http" && parsed.Scheme != "https" {
		return nil, ErrUsage("download URL must use HTTP or HTTPS scheme")
	}

	// Operation hooks
	op := OperationInfo{
		Service: "Account", Operation: "DownloadURL",
		ResourceType: "download", IsMutation: false,
	}
	if gater, ok := ac.parent.hooks.(GatingHooks); ok {
		if ctx, err = gater.OnOperationGate(ctx, op); err != nil {
			return
		}
	}
	start := time.Now()
	ctx = ac.parent.hooks.OnOperationStart(ctx, op)
	defer func() { ac.parent.hooks.OnOperationEnd(ctx, op, err, time.Since(start)) }()

	// URL rewriting: replace scheme+host with cfg.BaseURL origin, preserve path+query
	baseURL, baseErr := url.Parse(ac.parent.cfg.BaseURL)
	if baseErr != nil {
		return nil, fmt.Errorf("invalid base URL: %w", baseErr)
	}
	rewritten := &url.URL{
		Scheme:   baseURL.Scheme,
		Host:     baseURL.Host,
		Path:     parsed.Path,
		RawQuery: parsed.RawQuery,
		Fragment: parsed.Fragment,
	}
	rewrittenURL := rewritten.String()

	// Hop 1: Authenticated API request (capture redirect)
	req, err := http.NewRequestWithContext(ctx, "GET", rewrittenURL, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	if authErr := ac.parent.authStrategy.Authenticate(ctx, req); authErr != nil {
		return nil, authErr
	}
	req.Header.Set("User-Agent", ac.parent.userAgent)

	apiClient := &http.Client{
		Transport: ac.parent.httpClient.Transport, // loggingTransport — fires hooks
		Timeout:   0,                              // no client-level timeout — body may be streamed on direct 2xx
		CheckRedirect: func(*http.Request, []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}

	ctx = contextWithAttempt(ctx, 1)
	req = req.WithContext(ctx)

	resp, err := apiClient.Do(req)
	if err != nil {
		return nil, ErrNetwork(err)
	}

	// Dispatch on response status
	switch {
	case resp.StatusCode == 301 || resp.StatusCode == 302 || resp.StatusCode == 303 ||
		resp.StatusCode == 307 || resp.StatusCode == 308:
		// Redirect — extract Location, close body, proceed to hop 2
		location := resp.Header.Get("Location")
		_ = resp.Body.Close()
		if location == "" {
			return nil, ErrAPI(resp.StatusCode, fmt.Sprintf("redirect %d with no Location header", resp.StatusCode))
		}
		// Resolve relative Location against the rewritten API URL
		resolvedLocation := resolveURL(rewrittenURL, location)

		// Hop 2: fetch from signed URL
		signedResp, signedErr := ac.parent.fetchSignedDownload(ctx, resolvedLocation) //nolint:bodyclose // body ownership transfers to caller via DownloadResult
		if signedErr != nil {
			return nil, signedErr
		}
		return &DownloadResult{
			Body:          signedResp.Body,
			ContentType:   signedResp.Header.Get("Content-Type"),
			ContentLength: signedResp.ContentLength,
			Filename:      filenameFromURL(rawURL),
		}, nil

	case resp.StatusCode >= 200 && resp.StatusCode < 300:
		// Direct download — no second hop
		return &DownloadResult{
			Body:          resp.Body,
			ContentType:   resp.Header.Get("Content-Type"),
			ContentLength: resp.ContentLength,
			Filename:      filenameFromURL(rawURL),
		}, nil

	default:
		body, _ := io.ReadAll(io.LimitReader(resp.Body, maxErrorMessageLen*2))
		_ = resp.Body.Close()
		return nil, checkResponse(resp, body)
	}
}

// filenameFromURL extracts a filename from the last path segment of a URL.
// Falls back to "download" if the URL is unparseable or has no path segments.
func filenameFromURL(rawURL string) string {
	u, err := url.Parse(rawURL)
	if err != nil {
		return "download"
	}
	base := path.Base(u.Path)
	if base == "" || base == "." || base == "/" {
		return "download"
	}
	unescaped, err := url.PathUnescape(base)
	if err != nil {
		return base
	}
	return unescaped
}

File: python/src/basecamp/download.py
from __future__ import annotations

from dataclasses import dataclass
from urllib.parse import unquote, urlparse, urlunparse

import httpx

from basecamp import _security
from basecamp.errors import ApiError, NetworkError, UsageError


@dataclass(frozen=True)
class DownloadResult:
    body: bytes
    content_type: str
    content_length: int
    filename: str


def filename_from_url(url: str) -> str:
    """Extract filename from URL path, URL-decoded."""
    path = urlparse(url).path
    name = unquote(path.rsplit("/", 1)[-1]) if "/" in path else ""
    return name or "download"


def _rewrite_url(raw_url: str, base_url: str) -> str:
    """Replace scheme+host with base_url origin, preserving base_url path prefix and download path+query+fragment."""
    parsed = urlparse(raw_url)
    base = urlparse(base_url)
    base_path = base.path.rstrip("/")
    new_path = base_path + parsed.path if base_path else parsed.path
    rewritten = parsed._replace(scheme=base.scheme, netloc=base.netloc, path=new_path)
    return urlunparse(rewritten)


def _parse_content_length(value: str | None) -> int:
    if not value:
        return -1
    try:
        parsed = int(value)
        return parsed if parsed >= 0 else -1
    except (ValueError, TypeError):
        return -1


def download_sync(raw_url: str, *, http_client, config) -> DownloadResult:
    """Perform sync download: URL rewrite → authenticated hop 1 → redirect → unauthenticated hop 2."""
    _validate_url(raw_url)
    rewritten_url = _rewrite_url(raw_url, config.base_url)

    response = http_client.get_no_retry(rewritten_url)

    if response.status_code in {301, 302, 303, 307, 308}:
        location = response.headers.get("Location") or response.headers.get("location")
        if not location:
            raise ApiError(f"redirect {response.status_code} with no Location header")
        resolved_url = _security.resolve_url(rewritten_url, location)
        signed_response = _fetch_signed(resolved_url, timeout=config.timeout)
        return DownloadResult(
            body=signed_response.content,
            content_type=signed_response.headers.get("content-type", ""),
            content_length=_parse_content_length(signed_response.headers.get("content-length")),
            filename=filename_from_url(raw_url),
        )

    if 200 <= response.status_code < 300:
        return DownloadResult(
            body=response.content,
            content_type=response.headers.get("content-type", ""),
            content_length=_parse_content_length(response.headers.get("content-length")),
            filename=filename_from_url(raw_url),
        )

    raise ApiError(f"Download failed with status {response.status_code}", http_status=response.status_code)


async def download_async(raw_url: str, *, http_client, config) -> DownloadResult:
    """Perform async download: URL rewrite → authenticated hop 1 → redirect → unauthenticated hop 2."""
    _validate_url(raw_url)
    rewritten_url = _rewrite_url(raw_url, config.base_url)

    response = await http_client.get_no_retry(rewritten_url)

    if response.status_code in {301, 302, 303, 307, 308}:
        location = response.headers.get("Location") or response.headers.get("location")
        if not location:
            raise ApiError(f"redirect {response.status_code} with no Location header")
        resolved_url = _security.resolve_url(rewritten_url, location)
        signed_response = await _fetch_signed_async(resolved_url, timeout=config.timeout)
        return DownloadResult(
            body=signed_response.content,
            content_type=signed_response.headers.get("content-type", ""),
            content_length=_parse_content_length(signed_response.headers.get("content-length")),
            filename=filename_from_url(raw_url),
        )

    if 200 <= response.status_code < 300:
        return DownloadResult(
            body=response.content,
            content_type=response.headers.get("content-type", ""),
            content_length=_parse_content_length(response.headers.get("content-length")),
            filename=filename_from_url(raw_url),
        )

    raise ApiError(f"Download failed with status {response.status_code}", http_status=response.status_code)


def _validate_url(raw_url: str) -> None:
    if not raw_url:
        raise UsageError("download URL is required")
    parsed = urlparse(raw_url)
    if not parsed.scheme or not parsed.netloc:
        raise UsageError("download URL must be an absolute URL")
    if parsed.scheme not in ("http", "https"):
        raise UsageError("download URL scheme must be http or https")


def _fetch_signed(url: str, *, timeout: float) -> httpx.Response:
    """Unauthenticated GET for signed download URL."""
    try:
        with httpx.Client(timeout=timeout, follow_redirects=True) as client:
            response = client.get(url)
        if response.status_code >= 400:
            raise ApiError(f"download failed with status {response.status_code}", http_status=response.status_code)
        return response
    except httpx.HTTPError as e:
        raise NetworkError(f"Download failed: {e}") from e


async def _fetch_signed_async(url: str, *, timeout: float) -> httpx.Response:
    """Async unauthenticated GET for signed download URL."""
    try:
        async with httpx.AsyncClient(timeout=timeout, follow_redirects=True) as client:
            response = await client.get(url)
        if response.status_code >= 400:
            raise ApiError(f"download failed with status {response.status_code}", http_status=response.status_code)
        return response
    except httpx.HTTPError as e:
        raise NetworkError(f"Download failed: {e}") from e
